src/main/kotlin/org/rust/lang/core/dfa/ControlFlowGraph.kt

```diff
@@ -84,13 +89,103 @@ class ControlFlowGraph private constructor(
             cfgBuilder.addContainedEdge(bodyExit, fnExit)
             cfgBuilder.addContainedEdge(fnExit, termination)
 
-            return ControlFlowGraph(owner, graph, body, regionScopeTree, entry, fnExit)
+            val unreachableElements = collectUnreachableElements(graph, entry)
+
+            return ControlFlowGraph(owner, graph, body, regionScopeTree, entry, fnExit, unreachableElements)
         }
-    }
 
-    // TODO: Implement dead code elimination
-    @Suppress("unused")
-    fun isNodeReachable(item: RsElement) = graph.depthFirstTraversal(entry).any { it.data.element == item }
+        /**
+         * Collects unreachable elements, i.e. elements cannot be reached by execution of [body].
+         *
+         * Only collects [RsExprStmt]s, [RsLetDecl]s and tail expressions, ignoring conditionally disabled.
+         */
+        private fun collectUnreachableElements(graph: CFGGraph, entry: CFGNode): Set<RsElement> {
+            /**
+             * In terms of our control-flow graph, a [RsElement]'s node is not reachable if it cannot not be fully executed,
+             * since [CFGBuilder] processes elements in post-order.
+             * But partially executed elements should not be treated as unreachable because the execution
+             * actually reaches them; only some parts of such elements should be treated as unreachable instead.
+             *
+             * So, first of all, collect all the unexecuted [RsElement]s (including partially executed)
+             */
+            val unexecutedElements = buildSet<RsElement> {
+                val fullyExecutedNodeIndices = graph.depthFirstTraversal(entry).map { it.index }.toSet()
+                graph.forEachNode { node ->
+                    if (node.index !in fullyExecutedNodeIndices) {
+                        val element = node.data.element ?: return@forEachNode
+                        add(element)
+                    }
+                }
+            }
+            val unexecutedStmts = unexecutedElements
+                .filterIsInstance<RsStmt>()
+            val unexecutedTailExprs = unexecutedElements
+                .filterIsInstance<RsBlock>()
+                .mapNotNull { block ->
+                    block.expr?.takeIf { it in unexecutedElements }
+                }
+
+            /**
+             * If [unexecuted] produces termination by itself
+             * (which means its inner expression type is [TyNever], see the corresponding checks below),
+             * it should be treated as unreachable only if the execution does not reaches the beginning of [unexecuted].
+             * The latter means that previous statement is not fully executed.
+             * This is needed in order not to treat the first, basically reachable, `return` as unreachable
+             */
+            fun isUnreachable(unexecuted: RsElement, innerExpr: RsExpr?): Boolean {
```

Comment:
Well, this is an interesting one. The type of the loop expression is `()`, i.e.:
```Rust
let _: () = loop {
    return;
    break;
};
let _: ! = loop {
    return;
};
```
So my heuristic `Unexecuted element is definitely unreachable if its inner expression does not produce termination` does not work here.

I'm not sure if this situation is common and how we should actually deal with it. I guess it is quite rare and related only to loops, so as a possible solution, we can extend our check from
```Kotlin
if (innerExpr != null && innerExpr.type !is TyNever) { ... }
```
to
```Kotlin
if (innerExpr != null && innerExpr.type !is TyNever && innerExpr !is RsLooplikeExpr) { ... }
```
but this might affect other cases of unreachable `RsLooplikeExpr`. On the other hand, if this situation is very rare, we can just ignore it for now. What do you think?

Reply:
I agree that this case is rare so we can ignore it for now

Btw can't we extend heuristic so that any parent of executed element should be not marked as unreachable? Looks like In this example `return` is executed, so `return;` and `loop` will not be marked as unreachable

Reply:
Do I understand correctly that you are talking about replacing the current `innerExpr` checking with something like this?
```Kotlin
if (unexecuted.children.any { it !in unexecutedElements }) {
    return false
}
```
If so, I'm afraid that would have an unreasonable negative impact on the algorithmic complexity. What do you think?

Reply:
> If so, I'm afraid that would have an unreasonable negative impact on the algorithmic complexity.

Ok, so lets indeed ignore such cases for now since they are rare